lib/mongo/server/connectable.rb

```diff
@@ -88,10 +88,11 @@ def ensure_connected
         ensure_same_process!
         connect!
         begin
-          yield socket
-        rescue Exception => e
-          disconnect!
-          raise e
+          res = yield socket
```

Comment:
Yes you can double check but `ensure` blocks are most probably executed inside-out.